rich-text-editor: add word/character count

nishanthkarthik opened the pull request:
![image](https://github.com/PrairieLearn/PrairieLearn/assets/7759435/176dc7e1-cd74-4195-8025-5d28198bc842)

---

Credits to @firasm for the initial impl

Fixes #10009

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.py:
```diff
@@ -69,6 +70,11 @@ def prepare(element_html, data):
             f'Invalid output format "{output_format}". Must be either "html" or "markdown".'
         )
 
+    counter = pl.get_string_attrib(element, "counter", "none")
+    valid_counters = ["character", "word", "none"]
+    if counter not in valid_counters:
+        raise Exception(f"Invalid {counter=}. Valid options are {valid_counters}")
+
```
Consider an enum. See #9955. Bonus points (though not necessary) if you also implement that for the output format.

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js:
```diff
@@ -94,4 +94,38 @@
+    }
+  }
+
+  Quill.register('modules/counter', Counter);
```
We already have, in this editor, a handler for the text-change event. I'm wondering if it may be simpler to just use that handler instead of creating a totally separate module for this functionality. This would make the code shorter and easier to maintain in the future.

nishanthkarthik replied:
Keeping the components separate could help us easily tweak the priorities of different change events - for extremely large inputs, we could theoretically debounce character/word count updates but keep the inputElement setter as-is because we don't want to lose any unsaved changes. Of course, it is not impossible (maybe slightly more difficult) when the change events are merged.

jonatanschroeder replied:
That is an interesting point, though I don't see how working with a module would make this easier to handle. AFAICT you can always have multiple handlers for the same event, if this is a concern of yours.

I see a module as an option when working with something that will be migrated across multiple instances of quill, including cases you don't have control over, which is not the case here. In this context a simple event handler is probably enough.

nishanthkarthik replied:
If I was using rx, I'd create one subscription per on-change handler and apply different debouncings although now that I think about it, I can manually create two different event streams from one handler. Not a big difference syntactically imo.

nishanthkarthik commented:
I'll rebase once we finish review

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.mustache:
```diff
@@ -14,5 +20,9 @@ $(function() {
 <div class="pl-rich-text-editor-container">
-  <div id="rte-{{editor_uuid}}" class="border mathjax_ignore"></div>
+    <div id="rte-{{editor_uuid}}" class="border mathjax_ignore"></div>
+    <div class="pl-rich-text-editor-counter-container
+        {{#counter_disabled}}pl-rich-text-editor-counter-container-disabled{{/counter_disabled}}">
+        <div id="rte-counter-{{editor_uuid}}"></div>
+    </div>
```
Instead of hiding it, just don't add it...
```suggestion
    {{^counter_disabled}}
    <div class="pl-rich-text-editor-counter-container">
        <div id="rte-counter-{{editor_uuid}}"></div>
    </div>
    {{/counter_disabled}}
```
This also removes the need for the display: none rule in CSS.

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.mustache:
```diff
@@ -4,6 +4,12 @@ $(function() {
         readOnly: {{read_only}},
         placeholder: '{{placeholder}}',
         format: '{{format}}',
+        modules: {
+            counter: {
+                container: '#rte-counter-{{editor_uuid}}',
+                unit: '{{counter}}',
+            }
+        },
```
Assuming this functionality moves out of a module: the container ID can be retrieved with the UUID that the code already has, so no need to add as an option. The unit could be an option directly (without hierarchy).

jonatanschroeder commented on docs/elements.md:
```diff
@@ -946,6 +946,7 @@ Provides an in-browser rich text editor, aimed mostly at manual grading essay-ty
 | `placeholder`        | string  | "Your answer here" | Text to be shown in the editor as a placeholder when there is no student input.                                                                                                                                                                                                                                                                         ...
 | `format`             | string  | `html`             | Format used to save the student's response. The element supports `html` and `markdown` formats. This format also affects how the source file name or inner HTML is interpreted.                                                                                                                                                                         ...
 | `markdown-shortcuts` | boolean | `true`             | Whether or not the editor accepts shortcuts based on markdown format (e.g., typing `_word_` causes the word to become italic).                                                                                                                                                                                                                          ...
+| `counter`            | string  | `none`             | Display a `word` or `character` count. This measure is inaccurate for unicode characters.                                                                                                                                                                                                                                                               ...
```
```suggestion
| `counter`            | `word`, `character` or `none`  | `none`             | Display a `word` or `character` count. This measure is inaccurate for unicode characters.                                                                                                                                                                                                                                                                                                                                                             |
```

This will require running `make format` to reformat the table. Again, bonus points if you do the same for `format`.

jonatanschroeder commented on exampleCourse/questions/element/richTextEditor/question.html:
```diff
@@ -52,6 +52,22 @@
   
   <hr>
 
+  <pl-question-panel>
+    <p>A character/word counter can be turned on by setting <code>counter="character"</code>/
+    <code>counter="word"</code> respectively. The default setting, <code>counter="none"</code>
+    disables the counter.</p>
+
+    <pl-rich-text-editor file-name="word-count-demo" source-file-name="woodchuck-initial.html" counter="word">
+    </pl-rich-text-editor>
+
+    <hr>
+
+    <pl-rich-text-editor file-name="character-count-demo" source-file-name="woodchuck-initial.html" counter="character">
+    </pl-rich-text-editor>
+  </pl-question-panel>
+
+  <hr>
+
```
Instead of two editors, consider adding one, then using the other counter option in combination with one of the other examples. Also, make sure the `file-name` has an extension, preferably html.

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.css:
```diff
@@ -6,8 +6,22 @@
+  border-style: solid;
+  border-color: #ccc;
+  border-width: 0 1px 1px;
+  color: #aaa;
```
This is too light for accessibility standards (https://webaim.org/resources/contrastchecker/)

nishanthkarthik replied:
Does PL have a standardized color palette? (I think bootstrap?)

jonatanschroeder replied:
If you can use bootstrap may be better. Maybe text-info?

nishanthkarthik replied:
text-info was blue, so I went with text-secondary

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js:
```diff
@@ -94,4 +94,38 @@
+        // Splitting empty text returns a non-empty array
+        return trimmed.length > 0 ? trimmed.split(/\s+/).length : 0;
+      } else if (this.options.unit === 'character') {
+        return text.length;
```
To handle Unicode characters:
```suggestion
        return [...text].length;
```

nishanthkarthik replied:
I am not sure if spread count is 100% accurate

```js
[..."😍🦶🏾"].length ==> 3
```

nishanthkarthik replied:
It's still better than byte count, so I think I'll use this.

jonatanschroeder replied:
> I am not sure if spread count is 100% accurate
> 
> ```js
> [..."😍🦶🏾"].length ==> 3
> ```

That's still technically 3 characters. If you hit backspace on the foot you get a different icon, since that foot is a combination of two characters.

nishanthkarthik replied:
At least on firefox, backspace deletes the foot. CLion on the other hand deletes the color and leaves the foot behind

jonatanschroeder reviewed (commented):
Thank you for your contribution. I'm happy to take over this change if you don't have the hours allotted for it, just let me know.

jonatanschroeder commented:
> I'll rebase once we finish review

No need for rebase, we merge with squash, so a simple merge with master is fine. Note that you may want to merge at some point soon, since there are two recent changes to pl-rich-text-editor.js that are not included in your PR.

nishanthkarthik commented:
addressed the review comments (mostly broken into fixups). I wasn't sure if I resolve the threads after fixups or you do, so I left them as they are :)

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js:
```diff
@@ -13,6 +13,35 @@
+        // Splitting empty text returns a non-empty array
+        return trimmed.length > 0 ? trimmed.split(/\s+/).length : 0;
+      } else if (this.unit === 'character') {
+        return [...text].length;
```
Adding a comment here would clarify why we're not using `text.length` instead.
```suggestion
        // Use a spread so that Unicode characters are counted instead of utf-16 code units
        return [...text].length;
```

jonatanschroeder commented on apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js:
```diff
@@ -47,6 +76,9 @@
 
     quill.setContents(quill.clipboard.convert({ html: contents }));
 
+    const getText = () => quill.getText();
+    const counter = options.counter === 'none' ? null : new Counter(options.counter, uuid, getText);
```
Nitpicking:
```suggestion
    const counter = options.counter === 'none' ? null : new Counter(options.counter, uuid, quill.getText);
```

nishanthkarthik replied:
The closure is necessary to bind `quill` - Inlining it breaks the call to getText (since this is unbound).

jonatanschroeder replied:
Ah, good point, I missed that.

jonatanschroeder commented on docs/elements.md:
```diff
@@ -937,15 +937,16 @@ Provides an in-browser rich text editor, aimed mostly at manual grading essay-ty
+| `source-file-name`   | string  | None                          | Name of the source file with existing content to be displayed in the editor. The format of this file must match the format specified in the `format` attribute.                                                                                                                                                                              ...
+| `directory`          | string  | special                       | Directory where the source file with existing code is to be found. Only useful if `source-file-name` is used. If it contains one of the special names `clientFilesCourse` or `serverFilesCourse`, then the source file name is read from the course's special directories, otherwise the directory is expected to be in the question's own d ...
+| `placeholder`        | string  | "Your answer here"            | Text to be shown in the editor as a placeholder when there is no student input.                                                                                                                                                                                                                                                              ...
+| `format`             | string  | `html` or `markdown`          | Format used to save the student's response. This format also affects how the source file name or inner HTML is interpreted.                                                                                                                                                                                                                  ...
+| `markdown-shortcuts` | boolean | `true`                        | Whether or not the editor accepts shortcuts based on markdown format (e.g., typing `_word_` causes the word to become italic).                                                                                                                                                                                                               ...
+| `counter`            | string  | `word`, `character` or `none` | Display a `word` or `character` count.                                                                                                                                                                                                                                                                                                       ...
```
Note that the second column is the type, and the third is the default value.
```suggestion
| `format`             | `html` or `markdown` | `html` | Format used to save the student's response. This format also affects how the source file name or inner HTML is interpreted.                                                                                                                                                                                                                                                                                                                           |
| `markdown-shortcuts` | boolean | `true`                        | Whether or not the editor accepts shortcuts based on markdown format (e.g., typing `_word_` causes the word to become italic).                                                                                                                                                                                                                                                                                                                        |
| `counter`            | `word`, `character` or `none` | `none` | Whether a word or character count should be displayed at the bottom of the editor.                                                                                                                                                                                                                                                                                                                                                                                                                |
```

Again, this will require `make format`.

jonatanschroeder commented on exampleCourse/questions/element/richTextEditor/question.html:
```diff
@@ -47,11 +47,22 @@
 
+  <pl-question-panel>
+    <p>A character or word counter can be turned on by setting <code>counter="character"</code> or
+    <code>counter="word"</code> respectively. The default setting, <code>counter="none"</code>
```
Nit:
```suggestion
    <code>counter="word"</code> respectively. The default setting, <code>counter="none"</code>,
```

jonatanschroeder reviewed (commented):
A few minor issues, but almost there. Also, make sure you merge with master as you may need to handle some merge conflicts (I assume they'll be auto-merged by Git, though).

nishanthkarthik commented:
Thanks for the quick review
